File: src-tauri/src/resources/resource_types.rs
//! Resource type definitions and utilities
//!
//! Provides a centralized enum for Kubernetes resource types with conversion utilities.

use serde::{Deserialize, Serialize};

/// Kubernetes resource types supported by the application
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
// Default serialization is PascalCase (matches Kind)
pub enum ResourceType {
    Pod,
    Deployment,
    StatefulSet,
    DaemonSet,
    Job,
    CronJob,
    ConfigMap,
    Secret,
    Service,
    Ingress,
    PersistentVolumeClaim,
    PersistentVolume,
    StorageClass,
    Endpoints,
    Node,
}

impl ResourceType {
    /// Parse resource type from string (Strict Kind matching)
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "Pod" => Some(Self::Pod),
            "Deployment" => Some(Self::Deployment),
            "StatefulSet" => Some(Self::StatefulSet),
            "DaemonSet" => Some(Self::DaemonSet),
            "Job" => Some(Self::Job),
            "CronJob" => Some(Self::CronJob),
            "ConfigMap" => Some(Self::ConfigMap),
            "Secret" => Some(Self::Secret),
            "Service" => Some(Self::Service),
            "Ingress" => Some(Self::Ingress),
            "PersistentVolumeClaim" => Some(Self::PersistentVolumeClaim),
            "PersistentVolume" => Some(Self::PersistentVolume),
            "StorageClass" => Some(Self::StorageClass),
            "Endpoints" => Some(Self::Endpoints),
            "Node" => Some(Self::Node),
            _ => None,
        }
    }

    /// Get the Kind name (CamelCase singular)
    pub fn kind(&self) -> &'static str {
        match self {
            Self::Pod => "Pod",
            Self::Deployment => "Deployment",
            Self::StatefulSet => "StatefulSet",
            Self::DaemonSet => "DaemonSet",
            Self::Job => "Job",
            Self::CronJob => "CronJob",
            Self::ConfigMap => "ConfigMap",
            Self::Secret => "Secret",
            Self::Service => "Service",
            Self::Ingress => "Ingress",
            Self::PersistentVolumeClaim => "PersistentVolumeClaim",
            Self::PersistentVolume => "PersistentVolume",
            Self::StorageClass => "StorageClass",
            Self::Endpoints => "Endpoints",
            Self::Node => "Node",
        }
    }

    /// Get the plural name (lowercase plural for API paths)
    pub fn plural(&self) -> &'static str {
        match self {
            Self::Pod => "pods",
            Self::Deployment => "deployments",
            Self::StatefulSet => "statefulsets",
            Self::DaemonSet => "daemonsets",
            Self::Job => "jobs",
            Self::CronJob => "cronjobs",
            Self::ConfigMap => "configmaps",
            Self::Secret => "secrets",
            Self::Service => "services",
            Self::Ingress => "ingresses",
            Self::PersistentVolumeClaim => "persistentvolumeclaims",
            Self::PersistentVolume => "persistentvolumes",
            Self::StorageClass => "storageclasses",
            Self::Endpoints => "endpoints",
            Self::Node => "nodes",
        }
    }

    /// Check if this is a cluster-scoped resource (not namespaced)
    pub fn is_cluster_scoped(&self) -> bool {
        matches!(self, Self::Node | Self::PersistentVolume)
    }

    /// Get all supported resource types
    pub fn all() -> &'static [ResourceType] {
        &[
            Self::Pod,
            Self::Deployment,
            Self::StatefulSet,
            Self::DaemonSet,
            Self::Job,
            Self::CronJob,
            Self::ConfigMap,
            Self::Secret,
            Self::Service,
            Self::Ingress,
            Self::PersistentVolumeClaim,
            Self::PersistentVolume,
            Self::StorageClass,
            Self::Endpoints,
            Self::Node,
        ]
    }
}

impl std::fmt::Display for ResourceType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_from_str() {
        assert_eq!(ResourceType::from_str("Pod"), Some(ResourceType::Pod));
        // Strict case - these should now fail (return None)
        assert_eq!(ResourceType::from_str("pods"), None);
        assert_eq!(ResourceType::from_str("pod"), None);
        assert_eq!(
            ResourceType::from_str("PersistentVolumeClaim"),
            Some(ResourceType::PersistentVolumeClaim)
        );
        assert_eq!(ResourceType::from_str("unknown"), None);
    }

    #[test]
    fn test_kind_and_plural() {
        assert_eq!(ResourceType::Pod.kind(), "Pod");
        assert_eq!(ResourceType::Pod.plural(), "pods");
        assert_eq!(ResourceType::Deployment.kind(), "Deployment");
        assert_eq!(ResourceType::Deployment.plural(), "deployments");
    }

    #[test]
    fn test_cluster_scoped() {
        assert!(ResourceType::Node.is_cluster_scoped());
        assert!(ResourceType::PersistentVolume.is_cluster_scoped());
        assert!(!ResourceType::Pod.is_cluster_scoped());
    }
}

File: src/lib/resource-types.ts
/**
 * Resource Type Constants
 *
 * Centralized definitions for Kubernetes resource types.
 * Uses CamelCase singular form (Kind) as the primary format.
 * Provides utility functions for conversion to other formats.
 */

/**
 * Kubernetes resource types (CamelCase singular - matches Kind field)
 */
export const ResourceType = {
    Pod: "Pod",
    Deployment: "Deployment",
    StatefulSet: "StatefulSet",
    DaemonSet: "DaemonSet",
    Job: "Job",
    CronJob: "CronJob",
    ConfigMap: "ConfigMap",
    Secret: "Secret",
    Service: "Service",
    Ingress: "Ingress",
    PersistentVolumeClaim: "PersistentVolumeClaim",
    PersistentVolume: "PersistentVolume",
    StorageClass: "StorageClass",
    Endpoints: "Endpoints",
    Node: "Node",
    Event: "Event",
    Namespace: "Namespace",
    CustomResourceDefinition: "CustomResourceDefinition",
} as const;

export type ResourceKind = (typeof ResourceType)[keyof typeof ResourceType];



/**
 * Mapping from Kind to lowercase plural API path format
 */
const kindToPlural: Record<ResourceKind, string> = {
    Pod: "pods",
    Deployment: "deployments",
    StatefulSet: "statefulsets",
    DaemonSet: "daemonsets",
    Job: "jobs",
    CronJob: "cronjobs",
    ConfigMap: "configmaps",
    Secret: "secrets",
    Service: "services",
    Ingress: "ingresses",
    PersistentVolumeClaim: "persistentvolumeclaims",
    PersistentVolume: "persistentvolumes",
    StorageClass: "storageclasses",
    Endpoints: "endpoints",
    Node: "nodes",
    Event: "events",
    Namespace: "namespaces",
    CustomResourceDefinition: "customresourcedefinitions",
};

/**
 * Mapping from lowercase plural to Kind
 */
const pluralToKind: Record<string, ResourceKind> = {
    pods: "Pod",
    deployments: "Deployment",
    statefulsets: "StatefulSet",
    daemonsets: "DaemonSet",
    jobs: "Job",
    cronjobs: "CronJob",
    configmaps: "ConfigMap",
    secrets: "Secret",
    services: "Service",
    ingresses: "Ingress",
    persistentvolumeclaims: "PersistentVolumeClaim",
    persistentvolumes: "PersistentVolume",
    storageclasses: "StorageClass",
    endpoints: "Endpoints",
    nodes: "Node",
    events: "Event",
    namespaces: "Namespace",
    customresourcedefinitions: "CustomResourceDefinition",
};

/**
 * Convert resource type to lowercase plural form (for API paths)
 * @example toPlural("Pod") → "pods"
 */
export function toPlural(resourceType: ResourceKind): string {
    return kindToPlural[resourceType];
}

/**
 * Convert any resource string to Kind format
 * @example toKind("pods") → "Pod", toKind("Pod") → "Pod"
 */
export function toKind(resourceType: string): ResourceKind | null {
    // Already in Kind format
    if (resourceType in kindToPlural) {
        return resourceType as ResourceKind;
    }
    // Lowercase format
    const lower = resourceType.toLowerCase();
    return pluralToKind[lower] ?? null;
}

/**
 * Check if a string is a valid ResourceType
 */
export function isResourceType(value: string): value is ResourceKind {
    return value in kindToPlural || value.toLowerCase() in pluralToKind;
}
